usr/man/man1/vpr.1
.TH VPR 1 "27 July 1983"
.UC 4
.SH NAME
vpr, vprm, vpq, vprint \- raster printer/plotter spooler
.SH SYNOPSIS
.B vpr
[
.B \-W
] [
.B \-l
] [
.B \-v
] [
.B \-t
[
.B \-1234
font ]
] [
.B \-w
] [
\fB\-w\fIwidth\fR
] [
.B \-m
] [ name ... ]
.br
.B vprm
[
id ...
] [
filename ...
] [
owner ...
]
.br
.B vpq
.br
.B vprint
[
.B \-W
]
file ...
.SH DESCRIPTION
.I Vpr 
causes the
named files
to be queued for printing or typeset simulation on one of the available
raster printer/plotters.
If no files are named, the standard input is read.
By default the input is assumed to be line printer-like text.
For very wide plotters, the input is run through the filter
.I /usr/lib/sidebyside
giving it an argument of
.B \-w106
which arranges it four pages adjacent with 90 column lines (the rest is for the
left margin).  Since there are 8 lines per inch in the default printer font,
.I vpr
thus produces 86 lines per page (the top and bottom lines are left blank).
.PP
The following options are available:
.TP 15
.B \-l
Print the input in a more literal manner.
Page breaks are not inserted, and most
control characters (except format effectors: \\n, \\f, etc.)
are printed
(many control characters print
special graphics not in the ASCII character set.)
Tab and underline processing is still done.
If this option is not given,
control characters which are not format effectors are ignored,
and page breaks are inserted after an appropriate number of lines
have been printed on a page.
.TP 15
.B \-W
Queues files for printing on a wide output device, if available.
Normally, files are queued for printing on a narrow output device.
.TP 15
.B \-1234
Specifies a font to be mounted on font position \fIi\fR.  The daemon
will construct a \fI.railmag\fR file referencing
\fI/usr/lib/vfont/name.size\fR.
.TP 15
.B \-m
Report by
.IR mail (1)
when printing is complete.
.TP
.B \-w
(Applicable only to wide output devices.)
Do not run the input through
sidebyside.
Such processing has been done already, or full (440 character) printer
width is desired.
.TP
\fB\-w\fIwidth\fR
Use width
.I width
rather than 90 for
.I sidebyside.
.TP
.B \-v
Use the filter
.I /usr/lib/vrast
to convert the vectors to raster. The named files must be a parameter and
vector file (in that order) created by
.IR plot (3X)
routines.
.TP
.B \-t
Use the filter
.I /usr/lib/vcat
to typeset the input on the printer/plotter.
The input must have been generated by
.IR troff (1)
run with the
.B \-t
option.
This is not normally run directly
to wide output devices, since it is wasteful to run only one page across.
The program
.IR vtroff (1)
is normally used and arranges,
using
.I vsort
for printing to occur four pages across, conserving paper.
.PP
.I Vprm
removes entries from the raster device queues.
The id, filename or owner should be that reported by
.I vpq.
All appropriate files will be removed.
Both queues are always searched.
The id of each file removed from the queue will be printed.
.PP
.I Vpq
prints the queues.
Each entry in the queue is printed showing the owner of the queue entry,
an identification number,
the size of the entry in characters,
and the file which is to be printed.
The
.I id
is useful for removing a specific entry from the printer queue using
.I vprm
.PP
.I Vprint
is a shell script which
.I pr's
a copy of each named file on one of the electrostatic printer/plotters.
The files are normally printed on a narrow device;
.B \-W
option causes them to be printed on a wide device.
.SH FILES
.ta 2i
/usr/spool/v?d/*	device spool areas
.br
/usr/lib/v?d	daemons
.br
/usr/lib/vpd	Versatec daemon
.br
/usr/lib/vpf	filter for printer simulation
.br
/usr/lib/*vcat	filter for typeset simulation
.br
/usr/lib/vrast	filter for plot
.br
/usr/lib/sidebyside	filter for wide output
.SH "SEE ALSO"
troff(1), vfont(5), vp(4), pti(1),
vtroff(1), plot(3X)
.SH BUGS
The 1's (one's) and l's (lower-case el's) in a Benson-Varian's
standard character set look very similar; caution is advised.
.PP
A versatec's hardware character set is rather ugly.
.I Vprint
should use one of the constant width fonts to produce prettier listings.


named/CHANGES
This is version 4.8.1 of bind.  It is a test release that includes
version 4.8 with fixes, asynchronous zone transfer and better reload
capabilities.  Although it is currently described as a test release,
it is believed to be reasonably stable and more usable than the currently-
released version, 4.8.  The changes of note are:

  o The asynchronous zone transfer code previously posted to the bind
    mailing list has been integrated, completed and tested.  There are
    a number of changes from the version posted, including fixes to
    allow top-level domains to work and a simplification of the timer
    code.

  o The code for reloading the server has been changed so that only
    primary zones master files that have changed are reloaded.  The
    cache and secondary zones are not flushed, and the sortlist, domain,
    etc. are reset to correspond to the boot file contents.

  o Several bugs have been fixed: the name "*" is not interpreted as
    a wildcard in cached zones, only in primary zones.  Secondary servers
    no longer decrement the time-to-live of records by the time since
    they verified the zone with the master; as a result, they never
    hand out nameserver referrals with too short a TTL to be usable.
    A bug was fixed that caused secondary servers with out-of-date
    zones to return empty answers between the actual expiration time
    and the next timeout.

There are several other bugs that have been reported but have not yet
been fixed.  In addition, the next regular release of named will
support negative caching, but this has not been integrated.

I would appreciate receiving feedback on this release; in particular,
problems (or lack of problems) when installing on various systems.
I attempted to update the SysV code when integrating the zone-transfer,
but haven't tested it.

The notes from version 4.8 follow.

----------
Welcome to version 4.8 of bind.

There have been several changes to the named boot file (/etc/named.boot)
of which you should be aware.  The "domain" line for each zone is no longer
needed, but one such line may still be used to specify a default domain
to be used for queries containing names with only a single component.
The term "suffixes", which was added in version 4.7alpha, has been removed.

The manual page on named (named.8) has been updated to reflect all
these changes.  Please read this and look at the example files
before installation.  You should also note the changes in the
resolver code to support non-fully-qualified addresses and per-user
host aliases.  See hostname(7) for an overview.  Two new routines
have been added to the resolver library since the last test release:
res_query formulates a query, sends it, waits for a response and does
preliminary error checking; res_search implements the search rules
of gethostbyname using res_query.

The MX lookup routine in sendmail has been modified to use res_search.  
Also, dn_skip takes an additional parameter and has been renamed
to dn_skipname.  While old sendmail binaries will work with the new
version of bind, because of these changes, it is desirable to install
new sendmail sources and recompile sendmail.  Do not rebuild sendmail 
from old sources.   The new sendmail is on ucbarpa.Berkeley.EDU for
anonymous FTP from pub/4.3/sendmail.MX.tar and pub/4.3/sendmail.MX.tar.Z.

There have been numerous changes to named, fixing most of the known
bugs that can be fixed without major structural changes in the server.
Several server configurations that failed before should now work.
Certain robustness problems have been fixed, in particular bounds-
checking when processing incoming packets.  Two changes have been made
in preparation for negative caching: SOA records are sent in the authority
section in negative responses with NXDOMAIN set, and a bug was fixed that
caused confusion and repeated requests if a response had no error, no answer
and an SOA in the authority section.  As such responses are already sent
by other servers, and will be sent by the next release of BIND, it is
important that all sites upgrade to this version as quickly as possible.

The root "hint" cache and cache file remain the largest problem area,
along with named's naivete in accepting bogus server's data.
These will be addressed in the next release, along with asynchronous
zone transfers, intelligent reloading of zone files, faster startup,
and caching of negative responses.

This version (4.8) will replace the last officially released version (4.5).
Version 4.5 has a serious bug that causes the generation of a continuous
stream of bogons to the root domain servers (bogus queries with the query
response bit set and possibly garbage for nsid and rcode).  It is imperative
that these versions of named be replaced as fast as possible.  We urge you to
field 4.8 quickly, for the sake of the root domain servers.

	Mike Karels
	Jean Wood
	bind@ucbarpa.Berkeley.EDU


usr/src/old/vpr/vpr.c
#include <signal.h>
#include <pagsiz.h>
#define	BUFSIZ	BSIZE

/*
 * 	vpr -- varian and versatec (as printers) spooler
 */

#define VAD		"/usr/lib/vad"
#define VPD		"/usr/lib/vpd"
char	VAtfname[] =	"/usr/spool/vad/tfaXXXXX";
char	VAcfname[] =	"/usr/spool/vad/cfaXXXXX";
char	VAdfname[] =	"/usr/spool/vad/dfaXXXXX";
char	VPtfname[] =	"/usr/spool/vpd/tfaXXXXX";
char	VPcfname[] =	"/usr/spool/vpd/cfaXXXXX";
char	VPdfname[] =	"/usr/spool/vpd/dfaXXXXX";
char	*tfname;
char	*cfname;
/* char	*lfname; */
char	*dfname;
int	wide;
int	literal;
int	nact;
int	tff;
int	mailflg;
char	person[10];
int	inchar;
int	maxrec	= 2000;
char	*width = "-w106";
int	troffit;
int     plotit;
char	*fonts[4];

main(argc, argv)
	int argc;
	char *argv[];
{
	register char *arg, *remote;
	int c, f, fv, flag;
	int out();

	umask(0);
	if (signal(SIGINT, SIG_IGN) == SIG_DFL)
		signal(SIGINT, out);
	if (signal(SIGQUIT, SIG_IGN) == SIG_DFL)
		signal(SIGQUIT, out);
	if (signal(SIGHUP, SIG_IGN) == SIG_DFL)
		signal(SIGHUP, out);
	if (signal(SIGTERM, SIG_IGN) == SIG_DFL)
		signal(SIGTERM, out);
	remote = "$	remote	**,onl";
	flag = 0;
	wide = 0;
	tfname = VAtfname;
	cfname = VAcfname;
	dfname = VAdfname;
	while (argc>1 && (arg = argv[1])[0]=='-') {
		if (arg[1] && arg[2]) {
			remote[12] = arg[1];
			remote[13] = arg[2];
			remote[14] = 0;
		} else switch (arg[1]) {

		case 'W':
			wide++;
			tfname = VPtfname;
			cfname = VPcfname;
			dfname = VPdfname;
			break;

		case '-':
			remote[12] = 'r';
			remote[13] = '1';
			remote[14] = '\0';
			break;

		case 'c':
			flag = '+';
			break;

		case 'r':
			flag = '-';
			break;

                case 'v':
                        plotit = 1;
                        width = 0;
                        break;

		case 'm':
			mailflg = 1;
			break;

		case 't':
			troffit = 1;
			width = 0;
			break;

		case '4':
		case '3':
		case '2':
		case '1':
			fonts[arg[1] - '1'] = argv[2];
			troffit = 1;
			argv++;
			argc--;
			break;

		case 'w':
			if (troffit)
				break;
			if (arg[2] == 0)
				width = 0;
			else
				width = arg;
			break;

		case 'l':
			literal++;	/* Pass control chars through. */
			break;
		}
		argc--;
		argv++;
	}
	pidfn();
	tff = nfile(tfname);
	if (!wide)		/* varian doesn't use sidebyside. */
		width = 0;
	ident();
	if(argc == 1)
		copy(0);
	while(--argc > 0) {
		arg = *++argv;
/*
		if(flag == '+')
			goto cf;
 * This may not work because the daemon runs as someone else, so don't bother
		if(*arg == '/' && flag != '-') {
			card(literal ? 'G' : 'F', arg);
			nact++;
			continue;
		}
		if(link(arg, lfname) < 0)
			goto cf;
		card(literal ? 'G' : 'F', lfname);
		card('U', lfname);
		lfname[inchar]++;
		nact++;
		goto df;
 */

	cf:
		f = open(arg, 0);
		if(f < 0) {
			printf("Cannot open %s\n", arg);
			if (plotit) {
				--argc;
				arg = *++argv;
			}
			continue;
		}
                if (plotit) {
 			if (--argc > 0) {
				arg = *++argv;
				fv = open(arg, 0);
				if (fv < 0) {
					printf("Cannot open %s\n", arg);
					close(f);
					continue;
				}
			}
			else {
				printf("Versaplot requires parm and vector file\n");
				close(f);
				continue;
			}
			copy(fv);
			close(fv);
		}
		copy(f);
		close(f);

	df:
		if(flag == '-') {
			f = unlink(arg);
			if(f < 0)
				printf("Cannot remove %s\n", arg);
		}
	}

	if(nact) {
		tfname[inchar]--;
		f = link(tfname, dfname);
		if(f < 0) {
			printf("Cannot rename %s\n", dfname);
			tfname[inchar]++;
			out();
		}
		unlink(tfname);
		if (wide)
			execl(VPD, "vpd", 0);
		else
			execl(VAD, "vad", 0);
		dfname[inchar]++;
		printf("Daemon doesn't exist\n");
		exit(0);
	}
	out();
}

copy(f)
int f;
{
	int ff, i, nr, nc;
	static char buf[BUFSIZ];
	int status;

	for (i = 0; i < 3; i++)
		if (fonts[i])
			card('1' + i, fonts[i]);
	if (troffit)
		card('T', cfname);
	else if (plotit)
		card('P', cfname);
	else
		card(literal ? 'G' : 'F', cfname);
	card('U', cfname);
	ff = nfile(cfname);
	nc = 0;
	nr = 0;
	if (width) {
		int pvec[2];
		pipe(pvec);
		i = fork();
		if (i < 0) {
			printf("No more processes\n");
			out();
		}
		if (i == 0) {
			if (f != 0) {
				close(0);
				dup(f);
			}
			close(1);
			dup(pvec[1]);
			close(pvec[0]);
			close(pvec[1]);
			execl("/usr/lib/sidebyside", "sidebyside", width, 0);
			perror("/usr/lib/sidebyside");
			exit(1);
		}
		close(pvec[1]);
		close(f);
		f = pvec[0];
	}
	while((i = read(f, buf, BUFSIZ)) > 0) {
		write(ff, buf, i);
		nc += i;
		if(nc >= BUFSIZ) {
			nc -= BUFSIZ;
			nr++;
			if(nr > maxrec) {
				printf("Copy file is too large\n");
				break;
			}
		}
	}
	close(ff);
	nact++;
	wait(&status);
}

card(c, s)
int c;
char s[];
{
	char *p1, *p2;
	static char buf[BUFSIZ];
	int col;

	p1 = buf;
	p2 = s;
	col = 0;
	*p1++ = c;
	while((c = *p2++) != '\0') {
		*p1++ = c;
		col++;
	}
	*p1++ = '\n';
	write(tff, buf, col+2);
}

ident()
{
	int c, n;
	register char *b1p, *pp, *b2p;
	static char b1[100], b2[100];

	b1p = b1;
	if(getpw(getuid(), b1p)) {
		b1p = "pdp::::m0000,m000:";
	}
	n = 0;
	b2p = b2;
	while(*b2p++ = "$	ident	"[n++]);
	b2p--;
	n = 5;
	while(--n) while(*b1p++ != ':');
	while((*b2p++ = *b1p++) != ':');
	b2p[-1] = ',';
	b1p = b1;
	pp = person;
	while((c = *b1p++) != ':') {
		*b2p++ = c;
		*pp++ = c;
	}
	*b2p++ = 0;
	*pp++ = 0;
	card('L', person);
	if (mailflg)
		card('M', person);
}

pidfn()
{
	register i, j, c;
	int s;
	int p;

	s = p = getpid();
	p &= 077777;
	i = 0;
	while(tfname[i] != 'X')
		i++;
	i += 4;
	for(j=0; j<5; j++) {
		c = (p%10) + '0';
		if(s<0 && j==4)
			c += 4;
		p /= 10;
		tfname[i] = c;
		cfname[i] = c;
/*
		lfname[i] = c;
*/
		dfname[i] = c;
		i--;
	}
	inchar = i;
}

nfile(name)
char *name;
{
	register f;

	f = creat(name, 0644);
	if(f < 0) {
		printf("Cannot create %s\n", name);
		out();
	}
	name[inchar]++;
	return(f);
}

out()
{
	register i;

	signal(SIGINT, SIG_IGN);
	signal(SIGQUIT, SIG_IGN);
	signal(SIGHUP, SIG_IGN);
	signal(SIGTERM, SIG_IGN);
	i = inchar;
	while(tfname[i] != 'a') {
		tfname[i]--;
		unlink(tfname);
	}
	while(cfname[i] != 'a') {
		cfname[i]--;
		unlink(cfname);
	}
/*
	while(lfname[i] != 'a') {
		lfname[i]--;
		unlink(lfname);
	}
*/
	while(dfname[i] != 'a') {
		dfname[i]--;
		unlink(dfname);
	}
	exit(0);
}
